Source/Core/DolphinQt/Debugger/CodeTraceDialog.cpp

```diff
@@ -0,0 +1,773 @@
+  tmp_trace.address = PC;
+
+  // Pull all register numbers out and store them. Limited to Reg0 if ps operation, as ps get
+  // too complicated to track easily.
+  std::regex regis("\\W([rfp]\\d+)[^r^f]*(?:([rf]\\d+))?[^r^f\\d]*(?:([rf]\\d+))?");
+  std::smatch match;
+
+  // ex: add r4, r5, r6 -> Reg0, Reg1, Reg2. Reg0 is always the target register.
+  if (std::regex_search(tmp, match, regis))
+  {
+    tmp_trace.reg0 = match.str(1);
+    if (match[2].matched)
+      tmp_trace.reg1 = match.str(2);
+    if (match[3].matched)
+      tmp_trace.reg2 = match.str(3);
+
+    // Get Memory Destination if load/store. The only instructions that start with L are load and
+    // load immediate li/lis (excluded).
+    if (tmp.compare(0, 2, "st") == 0 || tmp.compare(0, 5, "psq_s") == 0)
+    {
+      tmp_trace.memory_dest = PowerPC::debug_interface.GetMemoryAddressFromInstruction(tmp);
+      tmp_trace.is_store = true;
+    }
+    else if ((tmp.compare(0, 1, "l") == 0 && tmp.compare(1, 1, "i") != 0) ||
+             tmp.compare(0, 5, "psq_l") == 0)
+    {
+      tmp_trace.memory_dest = PowerPC::debug_interface.GetMemoryAddressFromInstruction(tmp);
+      tmp_trace.is_load = true;
+    }
+  }
+
+  m_code_trace.push_back(tmp_trace);
+}
+
+bool CodeTraceDialog::UpdateIterator(std::vector<CodeTrace>::iterator& begin_itr_return,
+                                     std::vector<CodeTrace>::iterator& end_itr_return)
+{
+  // If the range is changed for reprocessing, this will change the iterator accordingly.
+  auto begin_itr_temp = m_code_trace.begin();
+  auto end_itr_temp = m_code_trace.end();
+
+  if (m_change_range->isChecked())
+  {
+    bool good;
+    u32 start = m_bp1->currentText().toUInt(&good, 16);
+    if (!good)
+      start = m_bp1->currentData().toUInt(&good);
+    if (!good)
+      return false;
+
+    u32 end = m_bp2->currentText().toUInt(&good, 16);
+    if (!good)
+      end = m_bp2->currentData().toUInt(&good);
+    if (!good)
+      return false;
+
+    begin_itr_temp = find_if(m_code_trace.begin(), m_code_trace.end(),
+                             [start](const CodeTrace& t) { return t.address == start; });
+    end_itr_temp = find_if(m_code_trace.rbegin(), m_code_trace.rend(), [end](const CodeTrace& t) {
+                     return t.address == end;
+                   }).base();
+  }
+
+  if (begin_itr_temp == m_code_trace.end() || end_itr_temp == m_code_trace.begin())
+    return false;
+
+  begin_itr_return = begin_itr_temp;
+  end_itr_return = end_itr_temp;
+  return true;
+}
+
+void CodeTraceDialog::ForwardTrace()
+{
+  std::vector<std::string> exclude{"dc", "ic", "mt", "c", "fc"};
+  TraceOutput tmp_out;
+  bool first_hit = true;
+
+  auto begin_itr = m_code_trace.begin();
+  auto end_itr = m_code_trace.end();
+  if (UpdateIterator(begin_itr, end_itr) == false)
+  {
+    m_error_msg = tr("Change Range using invalid addresses.");
+    return;
+  }
+
+  for (auto instr = begin_itr; instr != end_itr; instr++)
```

Comment:
My only issue is when we use the change range option.  I don't think ranged-based for loops have that option.  I could put a "if not yet at start address -> continue" and "if at end address -> break" detection inside a ranged-based loop, if it's more desirable though.

Reply:
Since I can't use a ranged-based loop in Backtrace(), should I keep the ForwardTrace() loop using iterators, so they behave the same?

Would it be better to pass iterators for the range to ForwardTrace() and Backtrace() or to pass a std::optional<u32> start and end address if the range is changed then calculate the iterators inside those functions?  I need to calculate them in CodeTraceDialog regardless.

Reply:
If you really can't use ranged-based for loop or if it makes the code harder to grasp, feel free to use normal for loops. I'm fine with iterators passed as parameters as well.